src/gui/touchscreenlayout.cpp

```diff
@@ -0,0 +1,348 @@
+	offset.Y = (pos.Y - (position.Y * screensize.Y)) / button_size;
+}
+
+bool ButtonLayout::isButtonAllowed(touch_gui_button_id id)
+{
+	return id != joystick_off_id && id != joystick_bg_id && id != joystick_center_id &&
+			id != touch_gui_button_id_END;
+}
+
+bool ButtonLayout::isButtonRequired(touch_gui_button_id id)
+{
+	return id == overflow_id;
+}
+
+s32 ButtonLayout::getButtonSize(v2u32 screensize)
+{
+	return std::min(screensize.Y / 4.5f,
+			RenderingEngine::getDisplayDensity() * 65.0f *
+					g_settings->getFloat("hud_scaling"));
+}
+
+ButtonLayout ButtonLayout::getDefault()
+{
+	return {{
+		{jump_id, {
+			v2f(1.0f, 1.0f),
+			v2f(-1.0f, -0.5f),
+		}},
+		{sneak_id, {
+			v2f(1.0f, 1.0f),
+			v2f(-2.5f, -0.5f),
+		}},
+		{zoom_id, {
+			v2f(1.0f, 1.0f),
+			v2f(-0.75f, -3.5f),
+		}},
+		{aux1_id, {
+			v2f(1.0f, 1.0f),
+			v2f(-0.75f, -2.0f),
+		}},
+		{overflow_id, {
+			v2f(1.0f, 1.0f),
+			v2f(-0.75f, -5.0f),
+		}},
+	}};
+}
+
+ButtonLayout ButtonLayout::loadFromSettings()
+{
+	bool restored = false;
+	ButtonLayout layout;
+
+	std::string str = g_settings->get("touch_layout");
+	if (!str.empty()) {
+		std::istringstream iss(str);
+		try {
+			layout.deserializeJson(iss);
+			restored = true;
+		} catch (const Json::Exception &e) {
+			warningstream << "Could not parse touchscreen layout: " << e.what() << std::endl;
+		}
+	}
+
+	if (!restored)
+		return ButtonLayout::getDefault();
+
+	return layout;
+}
+
+std::unordered_map<touch_gui_button_id, irr_ptr<video::ITexture>> ButtonLayout::texture_cache;
+
+video::ITexture *ButtonLayout::getTexture(touch_gui_button_id btn, ISimpleTextureSource *tsrc)
+{
+	if (texture_cache.count(btn) > 0)
+		return texture_cache.at(btn).get();
+
+	video::ITexture *tex = tsrc->getTexture(button_image_names[btn]);
+	if (!tex)
+		// necessary in the mainmenu
+		tex = tsrc->getTexture(porting::path_share + "/textures/base/pack/" +
+				button_image_names[btn]);
+	irr_ptr<video::ITexture> ptr;
+	ptr.grab(tex);
+	texture_cache[btn] = ptr;
+	return tex;
+}
+
+void ButtonLayout::clearTextureCache()
+{
+	texture_cache.clear();
+}
+
+core::recti ButtonLayout::getRect(touch_gui_button_id btn,
+		v2u32 screensize, s32 button_size, ISimpleTextureSource *tsrc)
+{
+	const ButtonMeta &meta = layout.at(btn);
+	v2s32 pos = meta.getPos(screensize, button_size);
+
+	v2u32 orig_size = getTexture(btn, tsrc)->getOriginalSize();
+	v2s32 size((f32)orig_size.X / (f32)orig_size.Y * button_size, button_size);
+
+	return core::recti(pos - size / 2, core::dimension2di(size));
+}
+
+std::vector<touch_gui_button_id> ButtonLayout::getMissingButtons()
+{
+	std::vector<touch_gui_button_id> missing_buttons;
+	for (u8 i = 0; i < touch_gui_button_id_END; i++) {
+		touch_gui_button_id btn = (touch_gui_button_id)i;
+		if (isButtonAllowed(btn) && layout.count(btn) == 0)
+			missing_buttons.push_back(btn);
+	}
+	return missing_buttons;
+}
+
+void ButtonLayout::serializeJson(std::ostream &os) const
```

Comment:
FWIW: We already support groups in `minetest.conf`, which would be a "native" format. This here works too, I suppose - as long nobody ever needs to edit the setting manually (JSON without spaces, indents, or new lines).

I don't think this really needs to be changed - although that's only my opinion.

Reply:
JSON has the benefit that you could e.g. share a layout as a file or send it from server to client as well without too much additional work